modules/mastodon/index.js

```diff
@@ -0,0 +1,320 @@
+      } catch(ex) {
+        moduleConfig = defaultConfig();
+      }
+    }
+    writeConfig();
+  });
+
+  var userAgent = moduleConfig && moduleConfig.userAgent || 'EuIrcBot';
+  try {
+    var ver = require('../../package.json').version;
+    userAgent += '/' + ver;
+  } catch (ex) {
+    log.error('error getting package version: ' + ex);
+  }
+
+  var userAgentUrl = (moduleConfig && moduleConfig.userAgentUrl
+    || 'https://github.com/euank/EuIrcBot/blob/{commitish}/modules/mastodon/info.md');
+  if (userAgentUrl.includes('{commitish}')) {
+    var commitish = 'master';
+    try {
+      commitish = execSync('git rev-parse --short HEAD', {'encoding': 'utf8'}).trim();
+    } catch (ex) {
+      log.error('error determining deployed commit: ' + ex);
+    }
+    userAgentUrl = userAgentUrl.replace(/{commitish}/g, commitish);
+  }
+  userAgent += ' (+' + userAgentUrl + ')';
+
+  request = require('request').defaults({
+    headers: {'User-Agent': userAgent}
+  });
+};
+
+function handleUrl(url, reply) {
+  fetchResource(url, (error, record) => {
+    if (error) {
+      error.mastodonUrl = url;
+      log.info(error);
+      reply('error: ' + error.message);
+    } else {
+      formatRecord(record, (s) => reply.custom({ replaceNewlines: true }, s));
+    }
+  });
+
+  /*
+   * Instance autodiscovery: hit /api/v1/timelines/public and look for any new
+   * instance domains, then run those through forceCheckHostname to see if they
+   * ought to be added to our list.
+   *
+   * This will cause Pleroma instances to be added to the whitelist, even
+   * though they don't support ActivityPub objects, because they support the
+   * Mastodon client API.
+   */
+
+  robotRequest(
+    { url: urlMod.resolve(url, '/api/v1/timelines/public') },
+    (error, response, body) => {
+      if (error || response.statusCode != 200)
+        return;
+
+      try {
+        body = JSON.parse(body);
+      } catch (e) {
+        return;
+      }
+
+      arrayUniq(body.map((item) => urlMod.parse(item.url))).forEach((url) => {
+        if (moduleConfig.blocks && moduleConfig.blocks.includes(url.host))
+          return;
+        if (!moduleConfig.instances || !moduleConfig.instances.includes(url.host))
+          forceCheckHostname(url, () => {});
+      });
+    }
+  );
+}
+
+/*
+ * When a user calls "!masto [url]", both run and url get called. The first
+ * thing we do in each of these methods is check the domain against
+ * knownInstances; run only continues if the instance is unknown, and url only
+ * continues if the instance is known.
+ */
+
+module.exports.run = function(remainder, parts, reply, command, from, to, text, raw) {
+  url = urlMod.parse(remainder);
+  if (!checkHostname(url.host)) {
+    forceCheckHostname(url, (ok) => {
+      if (ok) {
+        handleUrl(url.href, reply);
+      } else {
+        reply('error: not a mastodon instance');
+      }
+    });
+  }
+};
+
+module.exports.url = function(url, reply) {
+  url = urlMod.parse(url);
```

Comment:
`try / catch` for url parsing

Reply:
You okay with just dropping these exceptions?

Reply:
Actually, my bad for that comment. I forgot you rewrote the `onurl` module to ensure it parses before calling this, so this can't throw. You're right, disregard my comment. 👍